govcd/cse/4.2.0/capiyaml_cluster.tmpl

```diff
@@ -0,0 +1,153 @@
+metadata:
+  name: "{{.ClusterName}}"
+  namespace: "{{.TargetNamespace}}"
+  labels:
+    cluster-role.tkg.tanzu.vmware.com/management: ""
+    tanzuKubernetesRelease: "{{.TkrVersion}}"
+    tkg.tanzu.vmware.com/cluster-name: "{{.ClusterName}}"
+  annotations:
+    osInfo: "ubuntu,20.04,amd64"
+    TKGVERSION: "{{.TkgVersion}}"
+spec:
+  clusterNetwork:
+    pods:
+      cidrBlocks:
+        - "{{.PodCidr}}"
+    serviceDomain: cluster.local
+    services:
+      cidrBlocks:
+        - "{{.ServiceCidr}}"
+  controlPlaneRef:
+    apiVersion: controlplane.cluster.x-k8s.io/v1beta1
+    kind: KubeadmControlPlane
+    name: "{{.ClusterName}}-control-plane-node-pool"
+    namespace: "{{.TargetNamespace}}"
+  infrastructureRef:
+    apiVersion: infrastructure.cluster.x-k8s.io/v1beta2
+    kind: VCDCluster
+    name: "{{.ClusterName}}"
+    namespace: "{{.TargetNamespace}}"
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  name: capi-user-credentials
+  namespace: {{.TargetNamespace}}
+type: Opaque
+data:
+  username: "{{.UsernameB64}}"
+  refreshToken: "{{.ApiTokenB64}}"
+---
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta2
+kind: VCDCluster
+metadata:
+  name: "{{.ClusterName}}"
+  namespace: "{{.TargetNamespace}}"
+spec:
+  site: "{{.VcdSite}}"
+  org: "{{.Org}}"
+  ovdc: "{{.OrgVdc}}"
+  ovdcNetwork: "{{.OrgVdcNetwork}}"
+  {{- if .ControlPlaneEndpoint}}
+  controlPlaneEndpoint:
+    host: "{{.ControlPlaneEndpoint}}"
+    port: 6443
+  {{- end}}
+  {{- if .VirtualIpSubnet}}
+  loadBalancerConfigSpec:
+    vipSubnet: "{{.VirtualIpSubnet}}"
+  {{- end}}
+  useAsManagementCluster: false
+  userContext:
+    secretRef:
+      name: capi-user-credentials
+      namespace: "{{.TargetNamespace}}"
+---
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta2
+kind: VCDMachineTemplate
+metadata:
+  name: "{{.ClusterName}}-control-plane-node-pool"
+  namespace: "{{.TargetNamespace}}"
+spec:
+  template:
+    spec:
+      catalog: "{{.Catalog}}"
+      template: "{{.VAppTemplate}}"
+      sizingPolicy: "{{.ControlPlaneSizingPolicy}}"
+      placementPolicy: "{{.ControlPlanePlacementPolicy}}"
+      storageProfile: "{{.ControlPlaneStorageProfile}}"
+      diskSize: {{.ControlPlaneDiskSize}}
+---
+apiVersion: controlplane.cluster.x-k8s.io/v1beta1
+kind: KubeadmControlPlane
+metadata:
+  name: "{{.ClusterName}}-control-plane-node-pool"
+  namespace: "{{.TargetNamespace}}"
+spec:
+  kubeadmConfigSpec:
+    preKubeadmCommands:
+      - mv /etc/ssl/certs/custom_certificate_*.crt /usr/local/share/ca-certificates && update-ca-certificates
+    clusterConfiguration:
+      apiServer:
+        certSANs:
+          - localhost
+          - 127.0.0.1
+      controllerManager:
+        extraArgs:
+          enable-hostpath-provisioner: "true"
+      dns:
+        imageRepository: "{{.ContainerRegistryUrl}}"
+        imageTag: "{{.DnsVersion}}"
+      etcd:
+        local:
+          imageRepository: "{{.ContainerRegistryUrl}}"
+          imageTag: "{{.EtcdVersion}}"
```

Comment:
Is Terraform going to provide a map of etcd and coredns versions for a given Kubernetes version? How does the user know these values? Are we storing any map internally that users can refer to?

Reply:
The map will be stored in this repo, these versions are stored in the JSON file **govcd/cse/tkg_versions.json**.

When the user specifies a given OVA, we check in code that it is indeed a TKG OVA, and we inspect the internals to retrieve the `VERSION` property from the Product Section data. With this version we inspect the map and get the corresponding versions (if the version is not in the JSON map, we return an error).

That way the user doesn't need to provide these details, they are retrieved automatically from the OVA, as long it is a TKG OVA.

This logic is in the function [getTkgVersionBundleFromVAppTemplate](https://github.com/adambarreiro/go-vcloud-director/blob/new-methods-cse/govcd/cse_util.go#L742)